### PR TITLE
[Core] Adding more BB methods to `SearchUtilities`

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Part of https://github.com/KratosMultiphysics/Kratos/pull/11224

This pull request introduces changes to the `SearchUtilities` class, specifically to the `PointIsInsideBoundingBox` method. The method is responsible for determining whether a given point is inside a bounding box.

The changes include three different versions of the method, each with its own purpose:

1. **PointIsInsideBoundingBox**:
   - Parameters:
     - `const BoundingBox<Point>& rBoundingBox`: The bounding box object containing the maximum and minimum points.
     - `const array_1d<double, 3>& rCoords`: The coordinates of the point to check.
   - Description:
     - The method checks if the `rCoords` point is inside the specified `rBoundingBox` by comparing the coordinates in each direction (x, y, and z) to the maximum and minimum points of the bounding box.
     - If the point is inside the bounding box, `true` is returned; otherwise, `false` is returned.

2. **PointIsInsideBoundingBoxWithTolerance**:
   - Parameters:
     - `const BoundingBox<Point>& rBoundingBox`: The bounding box object containing the maximum and minimum points.
     - `const array_1d<double, 3>& rCoords`: The coordinates of the point to check.
     - `const double Tolerance`: The tolerance value to be applied to the bounding box.
   - Description:
     - This version of the method includes an additional tolerance parameter.
     - The method first applies the tolerance to the maximum and minimum points of the bounding box.
     - After applying the tolerance, the method checks if the `rCoords` point is inside the adjusted bounding box, similar to the previous versions.
     - If the point is inside the bounding box, `true` is returned; otherwise, `false` is returned.

**🆕 Changelog**

- [Adding more BB methods to SearchUtilities](https://github.com/KratosMultiphysics/Kratos/commit/2c124c6142385a47069556d0b71de27fe5c3cc52)

### CODE REVIEW

File: kratos/utilities/search_utilities.h [hunk 1]
```diff
@@ -73,22 +74,40 @@ class SearchUtilities
...
+        );
+
+    /**
+     * @brief This method checks if a point is inside any bounding box of the global bounding boxes considering a certain tolerance
+     * @param rBoundingBox The bounding box
+     * @param rCoords The coordinates of the point
+     * @param Tolerance The tolerance
+     * @return True if the point is inside the bounding box
+     */
+    static bool PointIsInsideBoundingBoxWithTolerance(
```
Comment (matekelemen) [inline kratos/utilities/search_utilities.h]: the function's signature is different and it's pretty straightforward that the extra argument is a tolerance, so this function could just be an overload of `PointIsInsideBoundingBox`. No need to explain the argument in the function's name too.

File: kratos/utilities/search_utilities.cpp [hunk 2]
```diff
@@ -21,6 +21,72 @@
...
+    auto min_point = rBoundingBox.GetMinPoint();
+    
+    // Apply Tolerances (only in non zero BB cases)
+    const double epsilon = std::numeric_limits<double>::epsilon();
+    if (norm_2(max_point) > epsilon && norm_2(min_point) > epsilon) {
+        for (unsigned int i=0; i<3; ++i) {
+            max_point[i] += Tolerance;
+            min_point[i] -= Tolerance;
+        }
+    }
```
Comment (matekelemen) [inline kratos/utilities/search_utilities.cpp:61]: This makes no sense to me. Why is this check any better than just checking the norm of the input point?

Furthermore, the docs read `This method checks if a point is inside any bounding box of the global bounding boxes considering a certain tolerance`, but the tolerance is applied **only if** the bounding box is triple-degenerate and exactly on the origin.

File: kratos/utilities/search_utilities.h [hunk 3]
```diff
@@ -73,22 +74,40 @@ class SearchUtilities
...
-        if (rCoords[0] < rBoundingBox[0] && rCoords[0] > rBoundingBox[1])           // check x-direction
-            if (rCoords[1] < rBoundingBox[2] && rCoords[1] > rBoundingBox[3])       // check y-direction
-                if (rCoords[2] < rBoundingBox[4] && rCoords[2] > rBoundingBox[5])   // check z-direction
-                    return true;
-        return false;
-    }
+        );
+
+    /**
+     * @brief This method checks if a point is inside any bounding box of the global bounding boxes considering a certain tolerance
```
Comment (matekelemen) [inline kratos/utilities/search_utilities.h]: what's a _global bounding box_?

**Review by matekelemen — COMMENTED** [review]
this check is almost identical to what `IntervalUtility` does, except it checks 3 axes instead of 1. I suggest creating an implementation that both objects can use to reduce code duplication.

File: kratos/utilities/search_utilities.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/utilities/search_utilities.h]: huuuuum, true. I will think about changing it

File: kratos/utilities/search_utilities.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/utilities/search_utilities.h]: My bad, corrected

File: kratos/utilities/search_utilities.cpp (hunk 2 above)
Comment (loumalouomega) [reply kratos/utilities/search_utilities.cpp:61]: Corrected the doc

File: kratos/utilities/search_utilities.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/utilities/search_utilities.h]: Done

File: kratos/utilities/search_utilities.cpp [hunk 4]
```diff
@@ -21,6 +21,56 @@
...
+        if (rCoords[1] < max_point[1] && rCoords[1] > min_point[1])       // check y-direction
+            if (rCoords[2] < max_point[2] && rCoords[2] > min_point[2])   // check z-direction
+                return true;
+    return false;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
 void SearchUtilities::ComputeBoundingBoxesWithTolerance(
```
Comment (philbucher) [inline kratos/utilities/search_utilities.cpp:74]: changing this fct name (and the usages in the mappingapp) would be awesome, but I am also fine if you do it in a separate PR
:pray:

File: kratos/utilities/search_utilities.cpp [hunk 5]
```diff
@@ -21,6 +21,56 @@
...
+{
+    // Get the bounding box points
+    const auto& r_max_point = rBoundingBox.GetMaxPoint();
+    const auto& r_min_point = rBoundingBox.GetMinPoint();
+
+    // The Bounding Box check
+    if (rCoords[0] < r_max_point[0] && rCoords[0] > r_min_point[0])           // check x-direction
+        if (rCoords[1] < r_max_point[1] && rCoords[1] > r_min_point[1])       // check y-direction
+            if (rCoords[2] < r_max_point[2] && rCoords[2] > r_min_point[2])   // check z-direction
+                return true;
```
Comment (philbucher) [inline kratos/utilities/search_utilities.cpp:37]: maybe there would be a way to unify this logic, since it seems prone to copy-paste errors :thinking:

**Review by philbucher — APPROVED** [review]

File: kratos/utilities/search_utilities.cpp (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/search_utilities.cpp:74]: Not done in this PR, I changed the name of a new function

File: kratos/utilities/search_utilities.cpp (hunk 5 above)
Comment (loumalouomega) [reply kratos/utilities/search_utilities.cpp:37]: Yep, we could
